homeassistant/components/homeconnect/switch.py

```diff
@@ -0,0 +1,189 @@
+        try:
+            self.device.appliance.stop_program()
+        except HomeConnectError as err:
+            _LOGGER.error("Error while trying to stop program: %s", err)
+        self.async_entity_update()
+
+    def update(self):
+        """Update the switch's status."""
+        state = self.device.appliance.status.get("BSH.Common.Root.ActiveProgram", {})
+        if state.get("value", None) == self.program_name:
+            self._state = True
+        else:
+            self._state = False
+        _LOGGER.debug("Updated, new state: %s", self._state)
+
+
+def convert_to_snake(camel):
+    """Convert from CamelCase to snake_case.
+
+    Taken from https://stackoverflow.com/questions/1175208/elegant-python-function-to-convert-camelcase-to-snake-case
+    """
+    snake = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", camel)
+    return re.sub("([a-z0-9])([A-Z])", r"\1_\2", snake).lower()
+
+
+class HomeConnectPowerSwitch(HomeConnectEntity, SwitchDevice):
+    """Power switch class for Home Connect."""
+
+    def __init__(self, device):
+        """Inititialize the entity."""
+        super().__init__(device, device.appliance.name)
+        self._state = None
+
+    @property
+    def is_on(self):
+        """Return true if the switch is on."""
+        return bool(self._state)
+
+    def turn_on(self, **kwargs):
+        """Switch the device on."""
+        _LOGGER.debug("tried to switch on %s", self.name)
+        try:
+            self.device.appliance.set_setting(
+                "BSH.Common.Setting.PowerState", "BSH.Common.EnumType.PowerState.On"
+            )
+        except HomeConnectError as err:
+            _LOGGER.error("Error while trying to turn on device: %s", err)
+            self._state = False
+        self.async_entity_update()
+
+    def turn_off(self, **kwargs):
+        """Switch the device off."""
+        _LOGGER.debug("tried to switch off %s", self.name)
+        try:
+            self.device.appliance.set_setting(
+                "BSH.Common.Setting.PowerState", self.device.power_off_state
+            )
+        except HomeConnectError as err:
+            _LOGGER.error("Error while trying to turn off device: %s", err)
+            self._state = True
+        self.async_entity_update()
+
+    def update(self):
+        """Update the switch's status."""
+        if (
+            self.device.appliance.status.get("BSH.Common.Setting.PowerState", {}).get(
+                "value", None
+            )
+            == "BSH.Common.EnumType.PowerState.On"
+        ):
+            self._state = True
+        elif (
+            self.device.appliance.status.get("BSH.Common.Setting.PowerState", {}).get(
+                "value", None
+            )
+            == self.device.power_off_state
+        ):
+            self._state = False
+        elif self.device.appliance.status.get(
+            "BSH.Common.Status.OperationState", {}
+        ).get("value", None) in [
+            "BSH.Common.EnumType.OperationState.Ready",
```

Comment:
Can we extract the constant lists of strings to module level constants so we don't need to create them every update?

Reply:
You're absolute right, will do so.